File: 백준/Gold/2294. 동전 2/동전 2.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

#setrecursionlimit(10**9)
input = stdin.readline

n , k = map(int,input().split())

dp = [inf for _ in range(k+1) ]

coins = []

for _ in range(n):
    c =  int(input())
    coins.append(c)


dp[0] = 0

for coin in coins:
    for j in range(coin,k+1):
        dp[j] = min(dp[j],dp[j-coin]+1)
        


print( -1 if dp[k] == inf else dp[k] )
    

File: 백준/Gold/2294. 동전 2/동전 2.swift
let nk = readLine()!.split{$0 == " "}.map{Int($0)!}
let (n,k) = (nk[0],nk[1])

var coins: [Int] = []

var dp: [Int] = [Int](repeating: Int.max, count: 100001)

for _ in 0..<n {
    let coin = Int(readLine()!)!
    coins.append(coin)
}

dp[0] = 0


for coin in coins {
    for i in stride(from: coin, to: k+1, by: 1) {
        if dp[i-coin] == Int.max {
            continue
        }
        
        dp[i] = min(dp[i],dp[i-coin]+1)
    }
}


print(dp[k] == Int.max ? -1 : dp[k])
